apps/verticalhome/js/app.js

```diff
@@ -158,6 +196,187 @@
 
     },
 
+    /**
+     * Save all items(added, deleted, rearranged) from homescrenn
+     * deletedApps - store all unpinned items
+     * deleted items temporary marked by css class "hidden-item"
+     * no params
+     */
+
+    saveAllPinnedApps: function(){
+
+      var pinnedItems= app.pinnedAppsManager.items;
+      if(this.deletedApps.length>0){
+        for(var z = 0; z < this.deletedApps.length; z++){
+          app.itemStore.deletePinnedAppItem(this.deletedApps[z]);
+        }
+        this.deletedApps.splice(0,this.deletedApps.length);
+      }
+
+      for(var i = 0; i<pinnedItems.length; i++){
+
+        if(pinnedItems[i].element.classList.contains('hidden-item')){
+          pinnedItems.splice(i,1);
+        }
+
+        var obj = {};
+
+        if(pinnedItems[i].entry && pinnedItems[i].entry!=null){
+
+          var curEntry = pinnedItems[i].entry;
+          obj.name = curEntry.name;
+
+          obj.manifestURL = curEntry.manifestURL ||
+            curEntry.targetApp.manifestURL;
+
+          obj.icon = curEntry.icon || curEntry.icon.baseURI;
+
+          obj.index = pinnedItems[i].index;
+
+          app.itemStore.savePinnedAppItem(obj);
+
+        }
+      }
+
+      Toaster.showToast({
+        messageL10nId: 'changes-saved',
+        latency: 3000
+      });
+
+    },
+
+    pinToMainScreen: function(){
+      var moreAppsSelected = '#more-apps-screen .selected';
+      var selMoreApp = document.querySelector(moreAppsSelected);
+      var selMoreAppIcon = selMoreApp.getAttribute('data-test-icon');
+      var selMoreAppDataId = selMoreApp.getAttribute('data-identifier');
+      var selMoreAppTitle = selMoreApp.querySelector('span.title').textContent;
+      var maxDataIndex = app.pinnedAppsManager.items.length-1;
+
+      app.pinnedAppsManager.addEntry( maxDataIndex,
+                                      selMoreAppDataId,
+                                      selMoreAppIcon,
+                                      selMoreAppTitle);
+
+      Array.prototype.slice.call(
+        document.querySelector('#more-apps-screen #icons').children
+      );
+
+      app.itemStore.applicationSource.
+        removeIconFromGrid( selMoreAppDataId );
+
+      this.renderGrid();
+
+      if (selMoreApp.parentNode) {
+        selMoreApp.parentNode.removeChild(selMoreApp);
+      }
+
+      this.hideMoreApps();
+
+      Toaster.showToast({
+        messageL10nId: 'added-to-home-screen',
+        latency: 3000
+      });
+
+      var pinnedApps = app.pinnedAppsNavigation;
+      pinnedApps.refresh();
+      pinnedApps.reset();
+      pinnedApps.refresh();
```

Comment:
Why do we need first ``refresh()`` call? We are resetting them anyway, or am I missing something?